iXBRLViewerPlugin/viewer/src/js/factset.js

```diff
@@ -0,0 +1,84 @@
+// Copyright 2019 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export function FactSet(facts) {
+    this._facts = facts;
+}
+
+FactSet.prototype._allDimensions = function() {
+    var dims = {};
+    for (var i = 0; i < this._facts.length; i++) {
+        var dd = Object.keys(this._facts[i].dimensions());
+        for (var j = 0; j < dd.length; j++) {
+            dims[dd[j]] = 1;
+        }
+    }
+    return Object.keys(dims);
+}
+
+FactSet.prototype.minimallyUniqueLabel = function(fact) {
+    if (!this._minimallyUniqueLabels) {
+        var allLabels = {};
+        var allDims = ["c", "p"].concat(this._allDimensions());
+        /* Assemble a map of arrays of all aspect labels for all facts, in a
+         * consistent order */
+        for (var i = 0; i < this._facts.length; i++) {
+            var f = this._facts[i];
+            allLabels[f.id] = [];
+            for (var j = 0; j < allDims.length; j++) {
+                var dd = f.aspects()[allDims[j]];
+                allLabels[f.id].push(dd ? dd.valueLabel() : "");
+            }
+        }
+        /* Iterate each aspect label and compare that label across all facts in
+         * the set */
+        var uniqueLabels = {};
+        var haveEmptyLabels;
+        for (var j = 0; j < allDims.length + 2; j++) {
+            var labelMap = {};
+            for (var i = 0; i < this._facts.length; i++) {
+                labelMap[allLabels[this._facts[i].id][j]] = true;
+            }
+
+            /* We have at least two different labels, so include this label */
+            var uniqueLabelsByLabel = {};
+            haveEmptyLabels = false;
+            if (Object.keys(labelMap).length > 1) {
+                for (var i = 0; i < this._facts.length; i++) {
+                    var fid = this._facts[i].id;
+                    var l = allLabels[fid][j];
+                    uniqueLabels[fid] = (uniqueLabels[fid] === undefined ? l : uniqueLabels[fid] + ", " + l);
+                    if (uniqueLabels[fid] === "") {
+                        haveEmptyLabels = true;
```

Comment:
Should this be outside the loop?  What is the intended formatting if one value is empty but another is not?  If the concepts are all the same, will the duplicate concept not be included twice?

Reply:
I think this was correct.  

If all concepts are the same, "concept" won't be added to the label as part of establishing unique labels, so there's no risk of it being duplicated.  This could have occurred if one of the concepts didn't actual have a label, but I've now fixed it to fall back on QName if there's no label, so concepts will always have a label.

The case that this code is intended to guard against is where you have, say, Revenue (Total) and Revenue (Europe).  If total is a default dimension value, the unique label code will give you "" and "Europe" as labels.  In this case we add back Concept as a label to get "Revenue" and "Revenue, Europe".

I've refactored this in order to handle the case of feeding it fully duplicate facts, and I think the end result is cleaner anyway.  Labels now don't get added to "uniqueLabels" until they've actually got some content, and then we just count how many we've got at the end to check that each fact has one (32ae878).